src/devices/bus/heathzenith/h89/sigmasoft_parallel_port.cpp

```diff
@@ -0,0 +1,245 @@
+	m_io_lo_addr(*this),
+	m_io_hi_addr(*this),
+	m_window_lo_addr(*this),
+	m_window_hi_addr(*this),
+	m_ctrl_w(*this)
+{
+}
+
+void sigmasoft_parallel_port::video_mem_w(u8 val)
+{
+	m_video_mem_w(val);
+}
+
+void sigmasoft_parallel_port::io_lo_addr_w(u8 val)
+{
+	m_io_lo_addr(val);
+}
+
+void sigmasoft_parallel_port::io_hi_addr_w(u8 val)
+{
+	m_io_hi_addr(val);
+}
+
+void sigmasoft_parallel_port::window_lo_addr_w(u8 val)
+{
+	m_window_lo_addr(val);
+}
+
+void sigmasoft_parallel_port::window_hi_addr_w(u8 val)
+{
+	m_window_hi_addr(val);
+}
+
+void sigmasoft_parallel_port::ctrl_w(u8 val)
+{
+	m_ctrl_w(val);
+}
+
+void sigmasoft_parallel_port::write(u8 select_lines, u16 offset, u8 data)
+{
+	offset -= BASE_ADDR;
+	if (!(select_lines & h89bus_device::H89_IO) || offset >= 8)
+	{
+		return;
+	}
+
+	LOGFUNC("%s: reg: %d val: %d\n", FUNCNAME, offset, data);
+
+	switch (offset)
+	{
+	case 0:
+		video_mem_w(data);
+		break;
+	case 1:
+		io_lo_addr_w(data);
+		break;
+	case 2:
+		io_hi_addr_w(data);
+		break;
+	case 3:
+		window_lo_addr_w(data);
+		break;
+	case 4:
+		window_hi_addr_w(data);
+		break;
+	case 5:
+		ctrl_w(data);
+		break;
+	case 6:
+		// TODO - Centronics interface
+		break;
+	case 7:
+		// TODO - Centronics interface
+		break;
+	}
+}
+
+u8 sigmasoft_parallel_port::video_mem_r()
+{
+	// get video memory value from igc device
+	return m_video_mem_r();
+}
+
+u8 sigmasoft_parallel_port::ctrl_r()
+{
+	// get control register from igc device
+	return m_ctrl_r();
+}
+
+u8 sigmasoft_parallel_port::read(u8 select_lines, u16 offset)
+{
+	offset -= BASE_ADDR;
+	u8 value = 0x00;
+
+	if (!(select_lines & h89bus_device::H89_IO) || offset >= 8)
+	{
+		return value;
+	}
+
+	switch (offset)
+	{
+	case 0:
+		value = video_mem_r();
+		break;
+	case 1:
+		// TODO - Light Pen Low address
+		break;
+	case 2:
+		// TODO - Light Pen High address
+		break;
+	case 3:
+		// TODO - Left input device
+		break;
+	case 4:
+		// TODO - Right input device
+		break;
+	case 5:
+		// Control Register
+		value = ctrl_r();
+		break;
+	case 6:
+		// TODO - Centronics interface
+		break;
+	case 7:
+		// TODO - Centronics interface
+		break;
+	}
+
+	LOGFUNC("%s: reg: %d val: %d\n", FUNCNAME, offset, value);
+
+	return value;
+}
+
+void sigmasoft_parallel_port::device_start()
+{
+}
+
+void sigmasoft_parallel_port::device_add_mconfig(machine_config &config)
+{
+	// connect callbacks to TLB
+	ctrl_r_cb().set(":tlbc", FUNC(heath_tlb_connector::sigma_ctrl_r));
+	video_mem_r_cb().set(":tlbc", FUNC(heath_tlb_connector::sigma_video_mem_r));
+	video_mem_cb().set(":tlbc", FUNC(heath_tlb_connector::sigma_video_mem_w));
+	io_lo_cb().set(":tlbc", FUNC(heath_tlb_connector::sigma_io_lo_addr_w));
+	io_hi_cb().set(":tlbc", FUNC(heath_tlb_connector::sigma_io_hi_addr_w));
+	window_lo_cb().set(":tlbc", FUNC(heath_tlb_connector::sigma_window_lo_addr_w));
+	window_hi_cb().set(":tlbc", FUNC(heath_tlb_connector::sigma_window_hi_addr_w));
+	ctrl_cb().set(":tlbc", FUNC(heath_tlb_connector::sigma_ctrl_w));
```

Comment:
It would be better not to use absolute tags here.

Reply:
What is the alternative? Before both the card and TLBC was created in h89.cpp, and could be directly linked. But here, when the parallel card is created, it does not have direct access to the tlbc.

Reply:
There are relative tags.

Reply:
It’s still a device making assumptions about the surrounding system, which is fragile, and we’re trying to eliminate.  If these signals can pass through the slot connector, they need to go through the slot device somehow.

It seems these are being added faster than existing ones are fixed.  At this rate I think I’m going to need to come up with a way to trigger validation warnings for them.

Reply:
I received feedback from some other mame devs that resulted in my followup PR based on their suggestions - https://github.com/mamedev/mame/pull/13040 which you have also commented on. I guess we can have the conversation on that PR.